File: lev1/약수의 개수와 덧셈.py
def solution(left, right):
    # 1.5910107344388962
    #ans = 0
    #for n in range(left, right+1):
    #    cnt = 0
    #    for i in range(1, int(n**0.5)+1):
    #        if n % i == 0:
    #            cnt += 2 if n / i != i else 1
    #    ans += n if cnt % 2 == 0 else -n
    #return ans

    # code refactoring (R) - 0.20098081603646278
    return sum([-n if (n**0.5).is_integer() else n for n in range(left, right+1)])


print(solution(13, 17))
print(solution(24, 27))
print(solution(1, 3))


if __name__ == "__main__":
    from timeit import Timer
    query = [[13,17], [24,27], [1,3], [1,1000]]
    t = Timer(f"for t in {query}: solution(*t)", "from __main__ import solution")
    print(t.timeit(number=1000))

File: lev1/약수의_개수와_덧셈.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>


int sum_square_number(int x){
    return (x * (x + 1) * (2 * x + 1) / 6);
}

int solution(int left, int right) {
    // int ans = 0;
    // for(int i=left; i<=right; i++){
    //     bool is_div = false;
    //     // 2.643107
    //     // for(int j=1; j<=(int)sqrt(i)+1; j++){
    //     // code refactoring - 1.376362(without callback function (sqrt))
    //     for(int j=1; j*j<=i; j++){
    //         if(i%j==0 & j*j==i){
    //             is_div = true;
    //             break;
    //         }
    //     }
    //     ans += is_div ? -i : i;
    // }
    // return ans;
    
    // code refactoring - 0.001346 -> 0.000599 (약 55% 성능 증진)
    int ans = (right-left+1)*(left+right)/2; // sum left to right
    ans -= 2 * (sum_square_number(sqrt(right)) - sum_square_number(sqrt(left - 1)));
    return ans;
}


int main(){
    clock_t tic = clock();
    for (int i=0; i<10000; i++){
        solution(13, 17);
        solution(24, 27);
        solution(1, 1);
        solution(1,1000);
        solution(1,997);
    }
    clock_t toc = clock();
    printf("running time: %f\n", (double)(toc-tic) / CLOCKS_PER_SEC);
    return 0;
}
